File: 2121-find-if-path-exists-in-graph/2121-find-if-path-exists-in-graph.cpp
class Solution {
public:
  bool dfs(int node, int destination, vector<vector<int>>& graph, vector<bool>& visited) {
        if (node == destination) {
            return true;
        }
        visited[node] = true;
        for (int neighbor : graph[node]) {
            if (!visited[neighbor]) {
                if (dfs(neighbor, destination, graph, visited)) {
                    return true;
                }
            }
        }
        return false;
    }

    bool validPath(int n, vector<vector<int>>& edges, int source, int destination) {
        // Step 1: Build the graph using adjacency list
        vector<vector<int>> graph(n);
        for (auto& edge : edges) {
            int u = edge[0];
            int v = edge[1];
            graph[u].push_back(v);
            graph[v].push_back(u);
        }
        
        // Step 2: Perform DFS to find if there's a path from source to destination
        vector<bool> visited(n, false);
        return dfs(source, destination, graph, visited);
    }
};

File: 2121-find-if-path-exists-in-graph/2121-find-if-path-exists-in-graph.js
/**
 * @param {number} n
 * @param {number[][]} edges
 * @param {number} source
 * @param {number} destination
 * @return {boolean}
 */
var validPath = function (n, edges, start, dest) {
  const graph = createGraph(edges);
  return hasPath(graph, start, dest);
};

const createGraph = (edges) => {
  const graph = {};
  for (const edge of edges) {
    const [a, b] = edge;
    if (!(a in graph)) graph[a] = new Set();
    if (!(b in graph)) graph[b] = new Set();
    graph[a].add(b);
    graph[b].add(a);
  }
  return graph;
};

const hasPath = (graph, start, dest) => {
  const stack = [start];
  const visited = new Set();

  while (stack.length > 0) {
    const node = stack.pop();
    if (node == dest) return true;
    if (!visited.has(node)) {
      visited.add(node);
      for (const neighbour of graph[node]) {
        stack.push(neighbour);
      }
    }
  }
  return false;
};
